reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,311 @@
+            .mean()
+            .reset_index(drop=False)
+            .rename(columns={f'{self.col_item}_paired': self.col_item})
+        )
+        all_recommendations_df[self.col_item] = (
+            all_recommendations_df[self.col_item].astype('int64')
+        )
+
+        return all_recommendations_df
+
+    def recommend_k_items(self,
+                          test,
+                          remove_seen=False,
+                          train=None,
+                          top_k=None):
+        """Recommend top K items for all users which are in the test set
+
+        Args:
+            test (pd.DataFrame): users to test
+            top_k (int): number of top items to recommend
+            remove_seen (bool): flag to remove recommendations
+            that have been seen by user
+
+        Returns:
+            pd.DataFrame: top k recommendation items for each user
+        """
+
+        test_users = test[[self.col_user]].drop_duplicates()
+        all_recommendations_df = self.predict(test)
+
+        if remove_seen:
+            if train is None:
+                raise ValueError("Please provide the training data to "
+                                 "remove seen items!")
+            # copy the DataFrame to avoid modification of the input
+            select_columns = [self.col_user, self.col_item]
+            temp_df = train[select_columns].copy()
+
+            logger.info("De-duplicating the user-item counts")
+            temp_df = temp_df.drop_duplicates(
+                select_columns, keep="last"
+            )
+
+            seen_item_indicator = (
+                all_recommendations_df
+                .merge(temp_df,
+                       on=select_columns,
+                       how='left',
+                       indicator=True)
+            )
+            # filtering original recommendations to novel ones
+            recommendations_df = (
+                seen_item_indicator
+                .loc[seen_item_indicator['_merge'] == 'left_only']
+            )
+
+            recommendations_df.drop(columns=['_merge'], inplace=True)
+            # remove dataframe to save memory
+            del temp_df
+        else:
+            recommendations_df = all_recommendations_df
+
+        # now only keep top_k
+        if not top_k:
+            top_k = self.num_recos
+
+        topk_recommendations_df = get_top_k_items(recommendations_df,
+                                                  col_user=self.col_user,
+                                                  col_rating=self.col_score,
+                                                  k=top_k)
+
+        # Making sure we have a row for every test user even if null
+        final_k_recommendations = (
+            test_users
+            .merge(topk_recommendations_df,
+                   on=self.col_user,
+                   how='left')
+        )
+
+        return final_k_recommendations
+
+    def eval_map_at_k(self, df_true, df_pred):  # noqa: N803
```

Comment:
we have a map_at_k here: https://github.com/microsoft/recommenders/blob/main/reco_utils/evaluation/python_evaluation.py#L578 which is optimized, can you please use this one?

Reply:
I think you don't need to write a new MAP@k function, you may use the one that exists already.

Reply:
yes removed it.